theano/tensor/opt.py

```diff
@@ -5901,6 +5901,52 @@ def local_log_add(node):
                 return [ret]
 
 
+@register_stabilize
+@register_specialize
+@gof.local_optimizer([T.log])
+def local_log_sum_exp(node):
+    # log(sum_i(exp(x_i))) = x_max + log(sum_i(exp(x_i - x_max)))
+
+    if node.op != T.log:
+        return
+
+    sum_node = node.inputs[0].owner
+    # If the sum has keepdims=True, there might be a dimshuffle
+    if sum_node and isinstance(sum_node.op, T.DimShuffle):
+        sum_node = sum_node.inputs[0].owner
+
+    if not sum_node or not isinstance(sum_node.op, T.Sum):
+        return
+
+    exp_node, axis = sum_node.inputs[0].owner, sum_node.op.axis
+    if not exp_node and exp_node.op != T.exp:
```

Comment:
I think it could be matched in a more general way, something like `isinstance(exp_node.op, Elemwise) and isinstance(exp_node.op.scalar_op, scalar.Exp)`